tests/passes/static-timing/compile-while-static.futil

```diff
@@ -20,19 +20,24 @@ component main() -> () {
       do_add[done] = add_r.done;
     }
 
-    group cond<"static"=1> {
+    group cond0<"static"=1> {
       lt.right = 32'd5;
       lt.left = 32'd1;
-      lt_r.in = lt.out;
-      lt_r.write_en = 1'b1;
-      cond[done] = lt_r.out;
+      comb_reg.in = lt.out;
+      comb_reg.write_en = 1'd1;
+      cond0[done] = comb_reg.done ? 1'd1;
     }
-
   }
 
   control {
-    while lt_r.out with cond {
-      do_add;
+    @static(3) seq {
+      @static cond0;
+      @bound @static(2) while comb_reg.out {
```

Comment:
The bound is there because I generated these test cases by running the normal pre-opt passes, and infer-static-timing won't add a static annotation to while nodes without a bound annotation. It doesn't get used by this pass at all, but this is what the IR must look like for this pass to work.

Reply:
Got it, so this PR does not fix https://github.com/cucapra/calyx/issues/545. We'll leave it open for now.

Reply:
Maybe I don't understand this test in that case–how can you know that the outermost `seq` will take `3` cycles if the inner `while` is not a bounded loop. Said differently, the only way to generate an FSM for the whole program is by making sure that the loop runs exactly once.

Reply:
This does not close #545. Right now this pass doesn't use the bound in any way. It also doesn't use the timing numbers on static annotations, except for enables. When I typed this test up, I just typed @static. I could have typed @static(2), and the outer seq would have @static(5). But this will generate the same FSM either way, because like I said, the pass only uses the static timing on enables. It statically runs the whole body, and after that, has an extra state which either transitions to the loop entry or the next state after the loop based on the condition. This adds extra latency to each iteration, but should be general. I think optimizing that is what #545 is about.

Reply:
For example, in the following control program:
```
    seq {
      cond0;
      @bound(4) while comb_reg.out {
        seq {
          do_add;
          cond0;
        }
      }
      cond0;
    }
```
we're saying that the while loop runs exactly three times. This means that `do_add` should be triggered on cycles 2, 4, 6, 8. But the generated FSM is:
```
enables:
[0, 1)
  cond0[go] = 1'd1;
[1, 2)
  do_add[go] = comb_reg.out ? 1'd1;
[2, 3)
  cond0[go] = comb_reg.out ? 1'd1;
[5, 6)
  cond0[go] = 1'd1;
transitions:
(0)->(1)
  1'b1
(1)->(2)
  comb_reg.out
(1)->(4)
  !comb_reg.out
(2)->(3)
  comb_reg.out
(3)->(1)
  comb_reg.out
(3)->(4)
  !comb_reg.out
(4)->(5)
  1'b1
(5)->(6)
  1'b1
```
Am I understanding the intended implementation of static while? It only makes sense to run a control program with `while` in it statically if the loop is bounded.

Reply:
>  It statically runs the whole body, and after that, has an extra state which either transitions to the loop entry or the next state after the loop based on the condition. This adds extra latency to each iteration, but should be general.

I see. Would that work correctly for the transitions after a loop? I think this kind of mixed dynamic-static compilation would be better left to `tdcc` instead of mixing it with `top-down-st`. Would switching out the implementation to just compile the `while` body and leaving the rest of the control program to `tdcc` work?

Reply:
Hmm yes probably. I can take a look at actually using the bound and implementing #545 shortly after this PR, or including it in this one if it's simple.

Reply:
Up to you! The options are disabling while compilation and leaving it up to tdcc or implementing bound.

Thinking about it more, disabling while compilation means that if the control program before and after the while is also fully static, the pass will now refuse to compile them unless they’re nested in their own seq block (which will not occur because collapse-control will remove such nestings). However, I think this is fine—we should have a separate pass that should collect together consecutive static control programs into seq blocks so that this pass compile them.

Reply:
One reason to implement bound handling separately is if we also fix #911, we can get rid of the conditions altogether. But again, that too can be separately addressed.

Reply:
I haven't thought about static parts before and after a while, but the one thing I'm actually interested in here is the body of a while that is static (with par nodes), since that's how I represent a pipeline. For now, I can remove the current logic for whiles and just let the visitor recurse into their bodies as if they didn't have the static annotation.

I think addressing #911 and the bound attribute makes a lot of sense. Either with tdcc or the top-down-st pass as it stands this morning, there are extra cycles of latency before, during, and after while loops. I think if we tackle #545 and #911 together, we can see clear improvements in the static path.

Reply:
Yup! That's what I was thinking too. The recursion in implicit in the visitor + the opportunistic compilation should handle static while bodies for now and make the pipeline work.

Reply:
I was just thinking about the static approach here a bit more, and I think it should be pretty simple to add now. Does this sound right?

If the bound attribute is present, compile it statically by recursing to the body `@bound` times in consecutive FSM states. If there is no bound, continue like we mentioned above so at least the body can potentially be handled statically.

Reply:
Yup! That sounds correct to me. The only implementation question is if you want to reset the fsm state to the first state in the body and have a separate counter (to avoid blowing up the main FSM) or just using the original FSM to count up.

Reply:
Ok, I will add that. I think it is working right now for the latter option.

What is the tradeoff for blowing up the main FSM? Wouldn't it be more bits of state to use separate counters? I'm trying to understand what that would gain.

Reply:
What I've been testing is in ebe100f. I tried it out on the pipeline I've been testing, and the trace looks great:

![Screenshot from 2022-02-21 12-27-50](https://user-images.githubusercontent.com/1182100/155016038-28f5c908-ee92-4af6-b867-4ccad2867f18.png)

Reply:
Awesome! So close to II=1 (need to fix multiplier to be truly pipelined to get that working).

Reply:
Is there an issue open for that? It would be great for the pipelined primitives to all be pipelined at the RTL level.

Reply:
Not yet! Feel free to open one

Reply:
Ok, we can take it here: https://github.com/cucapra/calyx/issues/928